### PR TITLE
issue_783

### LINKED ISSUE #783: JWT Middleware accepts expired tokens
Hi,

I can't rule out that I might be missing something in the config but I don't know what it is. Prestd (v1.2.2 in this case but so behaved v1.1.5 similarly) is working perfectly fine with expired JWT tokens too. So having a token is for life. The env vars I altered:

```
PREST_JWT_KEY=...
PREST_AUTH_TYPE=basic
PREST_DEBUG=false
PREST_AUTH_ENABLED=true
DATABASE_URL=postgres://...
```

Let me know if additional info needed. Any guidance would be greatly appreciated.

### PR DESCRIPTION
**Pull request by Listener430**
Added expired\notbefore validation for token claims for issue_783

Fix #783

### CODE REVIEW

**Comment by arxdsilva** [comment]
Please provide tests for the new function and middleware.

**Review by avelino — CHANGES_REQUESTED** [review]
Can you write test to validate the implementation?

**Review by avelino — APPROVED** [review]
LGTM

File: middlewares/middlewares_test.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	getToken := time.Now()
+	expireToken := time.Now().Add(time.Minute * 2)
+
+	// TODO: JWT any Algorithm support
+	sig, err := jose.NewSigner(
+		jose.SigningKey{
+			Algorithm: jose.HS256,
+			Key:       []byte(config.PrestConf.JWTKey)},
+		(&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
```
Comment (arxdsilva) [inline middlewares/middlewares_test.go:40]: Please use require here too

File: middlewares/middlewares_test.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+		return
+	}
+
+	cl := auth.Claims{
+		NotBefore: jwt.NewNumericDate(getToken),
+		Expiry:    jwt.NewNumericDate(expireToken),
+	}
+	bearer, _ := jwt.Signed(sig).Claims(cl).CompactSerialize()
```
Comment (arxdsilva) [inline middlewares/middlewares_test.go:48]: Don't forget to check here too with require

File: middlewares/middlewares_test.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	getToken := time.Now()
+	expireToken := time.Now().Add(-1 * time.Minute)
+
+	// TODO: JWT any Algorithm support
+	sig, err := jose.NewSigner(
+		jose.SigningKey{
+			Algorithm: jose.HS256,
+			Key:       []byte(config.PrestConf.JWTKey)},
+		(&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
```
Comment (arxdsilva) [inline middlewares/middlewares_test.go:80]: Same here about require

File: middlewares/middlewares_test.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		return
+	}
+
+	cl := auth.Claims{
+		NotBefore: jwt.NewNumericDate(getToken),
+		Expiry:    jwt.NewNumericDate(expireToken),
+	}
+	bearer, _ := jwt.Signed(sig).Claims(cl).CompactSerialize()
```
Comment (arxdsilva) [inline middlewares/middlewares_test.go:88]: Same here about checking the error

**Review by arxdsilva — CHANGES_REQUESTED** [review]

**Review by arxdsilva — APPROVED** [review]
We'll approve and adjust later, thanks!

**Pull request merged**